KOGITO-9971: [Guides] Dataindex deployment use cases with operator

nmirasch opened the pull request:
jira: https://issues.redhat.com/browse/KOGITO-9971

Please make sure that your PR meets the following requirements:

- [ ] You have read the [contributions doc](https://github.com/apache/incubator-kie-kogito-docs/blob/main/CONTRIBUTING.md)
- [ ] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [ ] Pull Request title contains the target branch if not targeting main: `[0.9.x] KOGITO-XYZ Subject`
- [ ] The nav.adoc file has a link to this guide in the proper category
- [ ] The index.adoc file has a card to this guide in the proper category, with a meaningful description

<details>
<summary>
How to backport a pull request to a different branch?
</summary>

In order to automatically create a **backporting pull request** please add one or more labels having the following format `backport-<branch-name>`, where `<branch-name>` is the name of the branch where the pull request must be backported to (e.g., `backport-7.67.x` to backport the original PR to the `7.67.x` branch).

> **NOTE**: **backporting** is an action aiming to move a change (usually a commit) from a branch (usually the main one) to another one, which is generally referring to a still maintained release branch. Keeping it simple: it is about to move a specific change or a set of them from one branch to another.

Once the original pull request is successfully merged, the automated action will create one backporting pull request per each label (with the previous format) that has been added.

If something goes wrong, the author will be notified and at this point a manual backporting is needed.

> **NOTE**: this automated backporting is triggered whenever a pull request on `main` branch is labeled or closed, but both conditions must be satisfied to get the new PR created.
</details>

tchughesiv commented:
does this use the new built-in data index deployment capabilities we've added to the operator? fyi, in case you hadn't seen or used it yet
https://github.com/apache/incubator-kie-kogito-docs/pull/518

nmirasch commented:
> does this use the new built-in data index deployment capabilities we've added to the operator? fyi, in case you hadn't seen or used it yet. #518
> 
> granted, not sure when a new release will occur that would make this feature available

This use cases were done before that new feature, so the answer is no. For the moment they are showing a way to see that deployments working, and I think those use cases will need to be updated including that new feature you commented and that will be reflected in the doc as well. @ricardozanini WDYT?

ricardozanini commented:
@nmirasch this is the reason why I mentioned you in #518 last week. Your guides should reflect that feature since we are going to release it together.

ricardozanini commented on serverlessworkflow/antora.yml:
```diff
@@ -109,6 +109,7 @@ asciidoc:
     kaoto_url: https://marketplace.visualstudio.com/items?itemName=redhat.vscode-kaoto
     minikube_url: https://minikube.sigs.k8s.io
     kogito_serverless_operator_url: https://github.com/apache/incubator-kie-kogito-serverless-operator/
+    flow_examples_operator_url: https://github.com/flows-examples/techpreview2/
```
We can't link to this example, it's our sandbox to use as reference only.

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/data-index/data-index-usecase-singleton.adoc:
```diff
@@ -0,0 +1,206 @@
+= Deploying Data Index and {product_name} application on Minikube
```
Same comment related to the #518 PR.

ricardozanini replied:
This comment is for the future, after merging this and refactor to incorporate Data Index on Operator guide.

ricardozanini commented:
Just to rectify my comment, we should merge this, then #518, and then open another to change this guide to link to #518. So we will have a version of this guide that won't include DI in the operator.

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/data-index/data-index-quarkus-extension.adoc:
```diff
@@ -39,12 +41,19 @@ These extensions are distributed as addons ready to work with different types of
+
+* kogito-addons-quarkus-data-index-persistence-postgresql
+* kogito-addons-quarkus-data-index-persistence-infinispan
+* kogito-addons-quarkus-data-index-persistence-mongodb
```
Can we add a little description of each?

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/data-index/data-index-quarkus-extension.adoc:
```diff
@@ -67,21 +79,39 @@ You can add the {data_index_ref} quarkus extension as an addon:
+
+[tabs]
+====
+Manually::
```
```suggestion
Manually to the POM.xml::
```

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/data-index/data-index-usecase-multi.adoc:
```diff
@@ -0,0 +1,221 @@
+* The `helloworld` workflow (no persistence), that is configured to register events to the Data Index Service.
+* Both workflows are configured to register the process events on the {data_index_ref} Service.
+
+You can directly access the UseCase2 example application we are going to follow at link:{flow_examples_operator_url}[{product_name} Use Cases repository].
```
```suggestion

```

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/data-index/data-index-usecase-multi.adoc:
```diff
@@ -0,0 +1,221 @@
+
+.Procedure
+. After cloning the link:{flow_examples_operator_url}[SonataFlow Use Cases repository]. Open a terminal and run the following commands
+. Create the namespace:
```
We can add the files to the guide inline directly instead. @kaldesai can we have an official OpenShift Serverless Logic examples/code repository?

domhanak replied:
For this requirement this is too late, let's do a kie-issue or requirement to tackle this.

domhanak replied:
:+1: for adding the directly here.

ricardozanini reviewed (commented):
Now we need to be creative and incorporate the use cases' sandbox repo here instead of linking to it.

nmirasch commented:
@ricardozanini removed the ref to usecases repo and put the dataindex deployment related kustomization resources in detail. While we are waiting for the change the doc to link the new feature about the dataindex deployment reference. could you take a look? Thanks!

domhanak commented on serverlessworkflow/modules/ROOT/pages/data-index/data-index-usecase-multi.adoc:
```diff
@@ -0,0 +1,198 @@
+
+
+
+.Use case kustomization.yaml resources that would deploy the both workflows inside usecases/usecase2 folder:
```
@nmirasch this procedure is missing steps, as a user I did not see any mention of me creating 

1. workflows (greeting, hello- world) - were not create by the user in the guide
2. service_discovery - was not created by the user in the guide
3. platforms/dataindex - platform was not created by user in the guide

domhanak replied:
Also the folder structure we reference here is something we need to fix and figure a way so user does not get into mess.

domhanak commented on serverlessworkflow/modules/ROOT/pages/data-index/data-index-usecase-singleton.adoc:
```diff
@@ -0,0 +1,183 @@
++
+--
+
+.Use case kustomization.yaml with resources that deploy the workflow inside usecases/usecase1 folder
```
Same a previous guide, I do not see these created by user, let's tell him to create these files too.

domhanak reviewed (changes requested):
There are some steps that are expecting something that does not exist, we need to include these in the guides.

domhanak commented on serverlessworkflow/modules/ROOT/pages/data-index/common/_dataindex_deployment_operator.adoc:
```diff
@@ -0,0 +1,218 @@
+* 01-postgres.yaml
+* 02-dataindex.yaml
+* application.properties
+
```
What about the `platform/` , `workflows` folders and the `service_discovery` are they needed? I notice them in output in commands, maybe an oversight? Not sure from my side, please confirm.

nmirasch commented:
I've added the referenced use cases to our official example here https://github.com/apache/incubator-kie-kogito-examples/pull/1839 and modified the guides links to that.

LuboTerifaj reviewed (commented):
Hello @nmirasch ,

I checked Data Index Quarkus extensions section (data-index-quarkus-extension.adoc file) and I have a few minor comments.

There is missing an explanation, what is the difference between "Quarkus Data Index" and "Quarkus Data Index persistence" extensions.

The referenced "PostgreSQL Configuration" section does not contain all necessary information about how to configure the database properly.
	
Thanks!

domhanak commented on serverlessworkflow/modules/ROOT/pages/data-index/data-index-quarkus-extension.adoc:
```diff
@@ -30,21 +32,32 @@ The example described in this document is based on the link:{kogito_sw_timeouts_
 The {data_index_ref} service has been designed to store and manage data from different workflow instances.
 Communication with the service is through events that contain the workflows related data and the service is responsible for storing them and exposing a GraphQL endpoint to allow queries and maintenance operations on the different workflow instances.
 
+image::data-index/data-index-addon.png[Image of data-index as a Quarkus Extension]
```
it is related to the paragraph above.

LuboTerifaj replied:
The paragraph above looks like more an overview of the Data Index service. The image shows the structure of the Workflow app, with and without persistence, without context. It's not a big issue, it was just my suggestion to improve documentation.

domhanak commented on serverlessworkflow/modules/ROOT/pages/data-index/data-index-quarkus-extension.adoc:
```diff
@@ -30,21 +32,32 @@ The example described in this document is based on the link:{kogito_sw_timeouts_
 Communication with the service is through events that contain the workflows related data and the service is responsible for storing them and exposing a GraphQL endpoint to allow queries and maintenance operations on the different workflow instances.
 
+image::data-index/data-index-addon.png[Image of data-index as a Quarkus Extension]
+
 In specific use cases, to avoid deploying the service separately, it could be useful to have the indexing functionality and the query capabilities embedded in the same application.
-For this purpose, the Quarkus {data_index_ref} extension can be added to any workflow application and incorporates the {data_index_ref} functionality into the same application without needing an external {data_index_ref} service.
+For this purpose, the Quarkus {data_index_ref} extension can be added to any workflow application and incorporates the full {data_index_ref} functionality into the same application without needing an external {data_index_ref} service.
 These extensions are distributed as addons ready to work with different types of persistence:
 
 * kogito-addons-quarkus-data-index-inmemory (inmemory PostgreSQL)
 * kogito-addons-quarkus-data-index-postgresql
 * kogito-addons-quarkus-data-index-infinispan
 * kogito-addons-quarkus-data-index-mongodb
 
+With the same purpose, the Quarkus {data_index_ref} persistence extension can be added to any workflow application and incorporates only the {data_index_ref} indexation and data persistence functionality into the same application without needing an external {data_index_ref} service to do that.
+These extensions are distributed as addons ready to work with different types of persistence:
+
+* kogito-addons-quarkus-data-index-persistence-postgresql
+* kogito-addons-quarkus-data-index-persistence-infinispan
+* kogito-addons-quarkus-data-index-persistence-mongodb
```
How?

LuboTerifaj replied:
There are duplicate parts of text, so it might worth to reformulate the whole section, but it is nothing critical.

domhanak commented on serverlessworkflow/modules/ROOT/pages/data-index/data-index-quarkus-extension.adoc:
```diff
@@ -54,11 +67,14 @@ The {data_index_ref} addon distribution added to the workflow must match the wor
 There is no need to configure the events connection for this purpose.
-* A new GraphQL endpoint is added to perform queries and management operations
+
+Only when any of the `kogito-addons-quarkus-data-index` addons is added:
+
+* A new GraphQL endpoint is added to perform queries and management operations when `kogito-addons-quarkus-data-index` is added
```
What is the duplicate information?

LuboTerifaj replied:
It is basically related to https://github.com/apache/incubator-kie-kogito-docs/pull/519/files#r1462125722, but as I mentioned, it is nothing critical.

domhanak commented on serverlessworkflow/modules/ROOT/pages/data-index/data-index-quarkus-extension.adoc:
```diff
@@ -67,21 +83,39 @@ You can add the {data_index_ref} quarkus extension as an addon:
 
 .Prerequisites
 * Your workflow is running and has persistence enabled.
+* {data_index_ref} is using the same datasource to store indexed data
```
It is a prerequisite, those are done in different pages, if you read the docs from start you should know be able to understand.
